// graphql-compose-mongoose/node8/discriminators/__mocks__/characterModels.js
"use strict";

exports.__esModule = true;
exports.getCharacterModels = getCharacterModels;
exports.getCharacterModelClone = getCharacterModelClone;
exports.CharacterObject = void 0;

var _mongooseCommon = require("../../__mocks__/mongooseCommon");

var _droidSchema = require("./droidSchema");

var _personSchema = require("./personSchema");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const enumCharacterType = {
  PERSON: 'Person',
  DROID: 'Droid'
};
const CharacterObject = {
  _id: {
    type: String,
    default: () => new _mongooseCommon.Types.ObjectId()
  },
  name: String,
  type: {
    type: String,
    require: true,
    enum: Object.keys(enumCharacterType)
  },
  kind: {
    type: String,
    require: true,
    enum: Object.keys(enumCharacterType)
  },
  friends: [String],
  // another Character
  appearsIn: [String] // movie

};
exports.CharacterObject = CharacterObject;
const CharacterSchema = new _mongooseCommon.Schema(CharacterObject);
const ACharacterSchema = new _mongooseCommon.Schema(_objectSpread({}, CharacterObject));

function getCharacterModels(DKey) {
  CharacterSchema.set('discriminatorKey', DKey);
  const CharacterModel = _mongooseCommon.mongoose.models.Character ? _mongooseCommon.mongoose.models.Character : _mongooseCommon.mongoose.model('Character', CharacterSchema);
  const PersonModel = _mongooseCommon.mongoose.models[enumCharacterType.PERSON] ? _mongooseCommon.mongoose.models[enumCharacterType.PERSON] : CharacterModel.discriminator(enumCharacterType.PERSON, _personSchema.PersonSchema);
  const DroidModel = _mongooseCommon.mongoose.models[enumCharacterType.DROID] ? _mongooseCommon.mongoose.models[enumCharacterType.DROID] : CharacterModel.discriminator(enumCharacterType.DROID, _droidSchema.DroidSchema);
  return {
    CharacterModel,
    PersonModel,
    DroidModel
  };
}

function getCharacterModelClone() {
  const NoDKeyCharacterModel = _mongooseCommon.mongoose.model('NoDKeyCharacter', ACharacterSchema);
  /*
    const APersonModel = ACharacterModel.discriminator('A' + enumCharacterType.PERSON, PersonSchema.clone());
     const ADroidModel = ACharacterModel.discriminator('A' + enumCharacterType.DROID, DroidSchema.clone());
  */


  return {
    NoDKeyCharacterModel
  }; // APersonModel, ADroidModel };
}
