// graphql-compose-mongoose/mjs/discriminators/__mocks__/characterModels.mjs
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { mongoose, Schema, Types } from '../../__mocks__/mongooseCommon';
import { DroidSchema } from './droidSchema';
import { PersonSchema } from './personSchema';
const enumCharacterType = {
  PERSON: 'Person',
  DROID: 'Droid'
};
export const CharacterObject = {
  _id: {
    type: String,
    default: () => new Types.ObjectId()
  },
  name: String,
  type: {
    type: String,
    require: true,
    enum: Object.keys(enumCharacterType)
  },
  kind: {
    type: String,
    require: true,
    enum: Object.keys(enumCharacterType)
  },
  friends: [String],
  // another Character
  appearsIn: [String] // movie

};
const CharacterSchema = new Schema(CharacterObject);
const ACharacterSchema = new Schema(_objectSpread({}, CharacterObject));
export function getCharacterModels(DKey) {
  CharacterSchema.set('discriminatorKey', DKey);
  const CharacterModel = mongoose.models.Character ? mongoose.models.Character : mongoose.model('Character', CharacterSchema);
  const PersonModel = mongoose.models[enumCharacterType.PERSON] ? mongoose.models[enumCharacterType.PERSON] : CharacterModel.discriminator(enumCharacterType.PERSON, PersonSchema);
  const DroidModel = mongoose.models[enumCharacterType.DROID] ? mongoose.models[enumCharacterType.DROID] : CharacterModel.discriminator(enumCharacterType.DROID, DroidSchema);
  return {
    CharacterModel,
    PersonModel,
    DroidModel
  };
}
export function getCharacterModelClone() {
  const NoDKeyCharacterModel = mongoose.model('NoDKeyCharacter', ACharacterSchema);
  /*
    const APersonModel = ACharacterModel.discriminator('A' + enumCharacterType.PERSON, PersonSchema.clone());
     const ADroidModel = ACharacterModel.discriminator('A' + enumCharacterType.DROID, DroidSchema.clone());
  */

  return {
    NoDKeyCharacterModel
  }; // APersonModel, ADroidModel };
}
