// graphql-compose-mongoose/es/resolvers/findByIds.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { limitHelper, limitHelperArgs, sortHelper, sortHelperArgs, projectionHelper } from './helpers';
import { beforeQueryHelper } from './helpers/beforeQueryHelper';
export default function findByIds(model, // === MongooseModel
tc, opts) {
  if (!model || !model.modelName || !model.schema) {
    throw new Error('First arg for Resolver findByIds() should be instance of Mongoose Model.');
  }

  if (!tc || tc.constructor.name !== 'ObjectTypeComposer') {
    throw new Error('Second arg for Resolver findByIds() should be instance of ObjectTypeComposer.');
  }

  return tc.schemaComposer.createResolver({
    type: [tc],
    name: 'findByIds',
    kind: 'query',
    args: _objectSpread({
      _ids: '[MongoID]!'
    }, limitHelperArgs(_objectSpread({}, opts && opts.limit)), {}, sortHelperArgs(tc, model, _objectSpread({
      sortTypeName: `SortFindByIds${tc.getTypeName()}Input`
    }, opts && opts.sort))),
    resolve: resolveParams => {
      const args = resolveParams.args || {};

      if (!Array.isArray(args._ids) || args._ids.length === 0) {
        return Promise.resolve([]);
      }

      const selector = {
        _id: {
          $in: args._ids
        }
      };
      resolveParams.query = model.find(selector); // eslint-disable-line

      resolveParams.model = model; // eslint-disable-line

      projectionHelper(resolveParams);
      limitHelper(resolveParams);
      sortHelper(resolveParams);
      return beforeQueryHelper(resolveParams);
    }
  });
}
