// graphql-compose/lib/utils/schemaPrinter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.printSchema = printSchema;
exports.printIntrospectionSchema = printIntrospectionSchema;
exports.isDefinedType = isDefinedType;
exports.printFilteredSchema = printFilteredSchema;
exports.printSchemaDefinition = printSchemaDefinition;
exports.isSchemaOfCommonNames = isSchemaOfCommonNames;
exports.printType = printType;
exports.printScalar = printScalar;
exports.printImplementedInterfaces = printImplementedInterfaces;
exports.printObject = printObject;
exports.printInterface = printInterface;
exports.printUnion = printUnion;
exports.printEnum = printEnum;
exports.printInputObject = printInputObject;
exports.printFields = printFields;
exports.printBlock = printBlock;
exports.printArgs = printArgs;
exports.printInputValue = printInputValue;
exports.printDirective = printDirective;
exports.printDeprecated = printDeprecated;
exports.printDescription = printDescription;
exports.printDescriptionWithComments = printDescriptionWithComments;

var _objectValues = _interopRequireDefault(require("graphql/polyfills/objectValues"));

var _inspect = _interopRequireDefault(require("graphql/jsutils/inspect"));

var _invariant = _interopRequireDefault(require("graphql/jsutils/invariant"));

var _printer = require("graphql/language/printer");

var _blockString = require("graphql/language/blockString");

var _introspection = require("graphql/type/introspection");

var _scalars = require("graphql/type/scalars");

var _directives = require("graphql/type/directives");

var _definition = require("graphql/type/definition");

var _astFromValue = require("graphql/utilities/astFromValue");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable consistent-return */
// copied from https://github.com/graphql/graphql-js/blob/master/src/utilities/schemaPrinter.js
// just expose all methods via adding export

/**
 * Accepts options as a second argument:
 *
 *    - commentDescriptions:
 *        Provide true to use preceding comments as the description.
 *
 */
function printSchema(schema, options) {
  return printFilteredSchema(schema, n => !(0, _directives.isSpecifiedDirective)(n), isDefinedType, options);
}

function printIntrospectionSchema(schema, options) {
  return printFilteredSchema(schema, _directives.isSpecifiedDirective, _introspection.isIntrospectionType, options);
}

function isDefinedType(type) {
  return !(0, _scalars.isSpecifiedScalarType)(type) && !(0, _introspection.isIntrospectionType)(type);
}

function printFilteredSchema(schema, directiveFilter, typeFilter, options) {
  const directives = schema.getDirectives().filter(directiveFilter);
  const typeMap = schema.getTypeMap();
  const types = (0, _objectValues.default)(typeMap).sort((type1, type2) => type1.name.localeCompare(type2.name)).filter(typeFilter);
  return `${[printSchemaDefinition(schema)].concat(directives.map(directive => printDirective(directive, options)), types.map(type => printType(type, options))).filter(Boolean).join('\n\n')}\n`;
}

function printSchemaDefinition(schema) {
  if (isSchemaOfCommonNames(schema)) {
    return;
  }

  const operationTypes = [];
  const queryType = schema.getQueryType();

  if (queryType) {
    operationTypes.push(`  query: ${queryType.name}`);
  }

  const mutationType = schema.getMutationType();

  if (mutationType) {
    operationTypes.push(`  mutation: ${mutationType.name}`);
  }

  const subscriptionType = schema.getSubscriptionType();

  if (subscriptionType) {
    operationTypes.push(`  subscription: ${subscriptionType.name}`);
  }

  return `schema {\n${operationTypes.join('\n')}\n}`;
}
/**
 * GraphQL schema define root types for each type of operation. These types are
 * the same as any other type and can be named in any manner, however there is
 * a common naming convention:
 *
 *   schema {
 *     query: Query
 *     mutation: Mutation
 *   }
 *
 * When using this naming convention, the schema description can be omitted.
 */


function isSchemaOfCommonNames(schema) {
  const queryType = schema.getQueryType();

  if (queryType && queryType.name !== 'Query') {
    return false;
  }

  const mutationType = schema.getMutationType();

  if (mutationType && mutationType.name !== 'Mutation') {
    return false;
  }

  const subscriptionType = schema.getSubscriptionType();

  if (subscriptionType && subscriptionType.name !== 'Subscription') {
    return false;
  }

  return true;
}

function printType(type, options) {
  if ((0, _definition.isScalarType)(type)) {
    return printScalar(type, options);
  } else if ((0, _definition.isObjectType)(type)) {
    return printObject(type, options);
  } else if ((0, _definition.isInterfaceType)(type)) {
    return printInterface(type, options);
  } else if ((0, _definition.isUnionType)(type)) {
    return printUnion(type, options);
  } else if ((0, _definition.isEnumType)(type)) {
    return printEnum(type, options);
  } else if ((0, _definition.isInputObjectType)(type)) {
    return printInputObject(type, options);
  } // Not reachable. All possible types have been considered.


  (0, _invariant.default)(false, `Unexpected type: ${(0, _inspect.default)(type)}`);
}

function printScalar(type, options) {
  return `${printDescription(type, options)}scalar ${type.name}`;
}

function printImplementedInterfaces(type) {
  const interfaces = type.getInterfaces ? type.getInterfaces() : [];
  return interfaces.length ? ` implements ${interfaces.map(i => i.name).join(' & ')}` : '';
}

function printObject(type, options) {
  return `${printDescription(type, options)}type ${type.name}${printImplementedInterfaces(type)}${printFields(type, options)}`;
}

function printInterface(type, options) {
  return `${printDescription(type, options)}interface ${type.name}${printImplementedInterfaces(type)}${printFields(type, options)}`;
}

function printUnion(type, options) {
  const types = type.getTypes();
  const possibleTypes = types.length ? ` = ${types.join(' | ')}` : '';
  return `${printDescription(type, options)}union ${type.name}${possibleTypes}`;
}

function printEnum(type, options) {
  const values = type.getValues().map((value, i) => `${printDescription(value, options, '  ', !i)}  ${value.name}${printDeprecated(value)}`);
  return `${printDescription(type, options)}enum ${type.name}${printBlock(values)}`;
}

function printInputObject(type, options) {
  const fields = (0, _objectValues.default)(type.getFields()).map((f, i) => `${printDescription(f, options, '  ', !i)}  ${printInputValue(f)}`);
  return `${printDescription(type, options)}input ${type.name}${printBlock(fields)}`;
}

function printFields(type, options) {
  const fields = (0, _objectValues.default)(type.getFields()).map((f, i) => `${printDescription(f, options, '  ', !i)}  ${f.name}${printArgs(f.args, options, '  ')}: ${String(f.type)}${printDeprecated(f)}`);
  return printBlock(fields);
}

function printBlock(items) {
  return items.length !== 0 ? ` {\n${items.join('\n')}\n}` : '';
}

function printArgs(args, options, indentation = '') {
  if (args.length === 0) {
    return '';
  } // If every arg does not have a description, print them on one line.


  if (args.every(arg => !arg.description)) {
    return `(${args.map(printInputValue).join(', ')})`;
  }

  return `(\n${args.map((arg, i) => `${printDescription(arg, options, `  ${indentation}`, !i)}  ${indentation}${printInputValue(arg)}`).join('\n')}\n${indentation})`;
}

function printInputValue(arg) {
  const defaultAST = (0, _astFromValue.astFromValue)(arg.defaultValue, arg.type);
  let argDecl = `${arg.name}: ${String(arg.type)}`;

  if (defaultAST) {
    argDecl += ` = ${(0, _printer.print)(defaultAST)}`;
  }

  return argDecl;
}

function printDirective(directive, options) {
  return `${printDescription(directive, options)}directive @${directive.name}${printArgs(directive.args, options)}${directive.isRepeatable ? ' repeatable' : ''} on ${directive.locations.join(' | ')}`;
}

function printDeprecated(fieldOrEnumVal) {
  if (!fieldOrEnumVal.isDeprecated) {
    return '';
  }

  const reason = fieldOrEnumVal.deprecationReason;
  const reasonAST = (0, _astFromValue.astFromValue)(reason, _scalars.GraphQLString);

  if (reasonAST && reason !== _directives.DEFAULT_DEPRECATION_REASON) {
    return ` @deprecated(reason: ${(0, _printer.print)(reasonAST)})`;
  }

  return ' @deprecated';
}

function printDescription(def, options, indentation = '', firstInBlock = true) {
  const {
    description
  } = def;

  if (description == null) {
    return '';
  }

  if (options && options.commentDescriptions) {
    return printDescriptionWithComments(description, indentation, firstInBlock);
  }

  const preferMultipleLines = description.length > 70;
  const blockString = (0, _blockString.printBlockString)(description, '', preferMultipleLines);
  const prefix = indentation && !firstInBlock ? `\n${indentation}` : indentation;
  return `${prefix + blockString.replace(/\n/g, `\n${indentation}`)}\n`;
}

function printDescriptionWithComments(description, indentation, firstInBlock) {
  const prefix = indentation && !firstInBlock ? '\n' : '';
  const comment = description.split('\n').map(line => indentation + (line !== '' ? `# ${line}` : '#')).join('\n');
  return `${prefix + comment}\n`;
}
