// graphql-compose/mjs/utils/configToDefine.mjs
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/* eslint-disable no-use-before-define */
// This is internal methods of graphql-js (introduced in 14.0.0)
// required for corret config convertion to internal field definition of types
// copy pasted from https://github.com/graphql/graphql-js/blame/master/src/type/definition.js
import invariant from 'graphql/jsutils/invariant';
import { inspect } from './misc';
import { isFunction, isObject } from './is';
import { SchemaComposer } from '../SchemaComposer';
import { ThunkComposer } from '../ThunkComposer';
import { ObjectTypeComposer } from '../ObjectTypeComposer';
import { InterfaceTypeComposer } from '../InterfaceTypeComposer';
import { getComposeTypeName } from './typeHelpers';

function isPlainObj(obj) {
  return obj && typeof obj === 'object' && !Array.isArray(obj);
}

export function defineFieldMap(config, fieldMap) {
  invariant(isPlainObj(fieldMap), `${config.name} fields must be an object with field names as keys or a ` + 'function which returns such an object.');
  const resultFieldMap = Object.create(null);

  for (const fieldName of Object.keys(fieldMap)) {
    const fieldConfig = fieldMap[fieldName];
    invariant(isPlainObj(fieldConfig), `${config.name}.${fieldName} field config must be an object`);
    invariant(!fieldConfig.hasOwnProperty('isDeprecated'), `${config.name}.${fieldName} should provide "deprecationReason" ` + 'instead of "isDeprecated".');

    const field = _objectSpread({}, fieldConfig, {
      isDeprecated: Boolean(fieldConfig.deprecationReason),
      name: fieldName
    });

    invariant(field.resolve == null || typeof field.resolve === 'function', `${config.name}.${fieldName} field resolver must be a function if ` + `provided, but got: ${inspect(field.resolve)}.`);
    const argsConfig = fieldConfig.args;

    if (!argsConfig) {
      field.args = [];
    } else {
      invariant(isPlainObj(argsConfig), `${config.name}.${fieldName} args must be an object with argument names as keys.`);
      field.args = Object.keys(argsConfig).map(argName => {
        const arg = argsConfig[argName];
        return {
          name: argName,
          description: arg.description === undefined ? null : arg.description,
          type: arg.type,
          defaultValue: arg.defaultValue,
          astNode: arg.astNode
        };
      });
    }

    resultFieldMap[fieldName] = field;
  }

  return resultFieldMap;
}
export function convertObjectFieldMapToConfig(fieldMap, schemaComposer) {
  const fields = {};
  const isThunk = isFunction(fieldMap);

  const _fields = isThunk ? fieldMap() : fieldMap;

  if (!isObject(_fields)) return {};
  Object.keys(_fields).forEach(n => {
    const _fields$n = _fields[n],
          {
      name,
      isDeprecated
    } = _fields$n,
          fc = _objectWithoutProperties(_fields$n, ["name", "isDeprecated"]);

    const args = {};

    if (Array.isArray(fc.args)) {
      // `fc.args` is an Array in `GraphQLFieldMap`
      fc.args.forEach(arg => {
        const {
          name: argName
        } = arg,
              ac = _objectWithoutProperties(arg, ["name"]);

        args[argName] = _objectSpread({}, ac, {
          type: isThunk ? new ThunkComposer(() => schemaComposer.typeMapper.convertInputTypeDefinition(ac.type || arg)) : schemaComposer.typeMapper.convertInputTypeDefinition(ac.type || arg)
        });
      });
      fc.args = args;
    } else if (isObject(fc.args)) {
      // `fc.args` is Object in `ObjectTypeComposerFieldConfigMapDefinition`
      Object.keys(fc.args).forEach(argName => {
        args[argName] = _objectSpread({}, isObject(fc.args[argName]) ? fc.args[argName] : null, {
          type: isThunk ? new ThunkComposer(() => schemaComposer.typeMapper.convertInputTypeDefinition(fc.args[argName].type || fc.args[argName])) : schemaComposer.typeMapper.convertInputTypeDefinition(fc.args[argName].type || fc.args[argName])
        });
      });
      fc.args = args;
    }

    fields[n] = _objectSpread({}, fc, {
      type: isThunk ? new ThunkComposer(() => schemaComposer.typeMapper.convertOutputTypeDefinition(fc.type || _fields[n])) : schemaComposer.typeMapper.convertOutputTypeDefinition(fc.type || _fields[n])
    });
  });
  return fields;
}
export function defineEnumValues(type, valueMap) {
  invariant(isPlainObj(valueMap), `${type.name} values must be an object with value names as keys.`);
  return Object.keys(valueMap).map(valueName => {
    const value = valueMap[valueName];
    invariant(isPlainObj(value), `${type.name}.${valueName} must refer to an object with a "value" key ` + `representing an internal value but got: ${inspect(value)}.`);
    invariant(!value.hasOwnProperty('isDeprecated'), `${type.name}.${valueName} should provide "deprecationReason" instead of "isDeprecated".`);
    return {
      name: valueName,
      description: value.description,
      isDeprecated: Boolean(value.deprecationReason),
      deprecationReason: value.deprecationReason,
      astNode: value.astNode,
      value: value.hasOwnProperty('value') ? value.value : valueName,
      extensions: undefined
    };
  });
}
export function defineInputFieldMap(config, fieldMap) {
  invariant(isPlainObj(fieldMap), `${config.name} fields must be an object with field names as keys or a ` + 'function which returns such an object.');
  const resultFieldMap = Object.create(null);

  for (const fieldName of Object.keys(fieldMap)) {
    const field = _objectSpread({}, fieldMap[fieldName], {
      name: fieldName
    });

    invariant(!field.hasOwnProperty('resolve'), `${config.name}.${fieldName} field has a resolve property, but ` + 'Input Types cannot define resolvers.');
    resultFieldMap[fieldName] = field;
  }

  return resultFieldMap;
}
export function convertInputFieldMapToConfig(fieldMap, schemaComposer) {
  const fields = {};
  const isThunk = isFunction(fieldMap);

  const _fields = isThunk ? fieldMap() : fieldMap;

  Object.keys(_fields).forEach(n => {
    const _fields$n2 = _fields[n],
          {
      name,
      isDeprecated
    } = _fields$n2,
          fc = _objectWithoutProperties(_fields$n2, ["name", "isDeprecated"]);

    fields[n] = _objectSpread({}, fc, {
      type: isThunk ? new ThunkComposer(() => schemaComposer.typeMapper.convertInputTypeDefinition(fc.type || _fields[n])) : schemaComposer.typeMapper.convertInputTypeDefinition(fc.type || _fields[n])
    });
  });
  return fields;
}
export function convertObjectTypeArrayAsThunk(types, sc) {
  const isThunk = isFunction(types);
  const t = isThunk ? types() : types;
  if (!Array.isArray(t)) return [];
  return t.map(type => {
    if (type instanceof ObjectTypeComposer || type instanceof ThunkComposer) {
      return type;
    }

    const tc = sc.typeMapper.convertOutputTypeDefinition(type);

    if (!tc && isThunk) {
      return new ThunkComposer(() => sc.typeMapper.convertOutputTypeDefinition(type), getComposeTypeName(type));
    }

    if (!(tc instanceof ObjectTypeComposer) && !(tc instanceof ThunkComposer)) {
      throw new Error(`Should be provided ObjectType but recieved ${inspect(type)}`);
    }

    return tc;
  });
}
export function convertInterfaceArrayAsThunk(types, sc) {
  const isThunk = isFunction(types);
  const t = isThunk ? types() : types;
  if (!Array.isArray(t)) return [];
  return t.map(type => {
    if (type instanceof InterfaceTypeComposer || type instanceof ThunkComposer) {
      return type;
    }

    return isThunk ? new ThunkComposer(() => sc.typeMapper.convertInterfaceTypeDefinition(type), getComposeTypeName(type)) : sc.typeMapper.convertInterfaceTypeDefinition(type);
  });
}
