// graphql-compose-mongoose/node8/resolvers/updateOne.js
"use strict";

exports.__esModule = true;
exports.default = updateOne;

var _helpers = require("./helpers");

var _findOne = _interopRequireDefault(require("./findOne"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function updateOne(model, // === MongooseModel
tc, opts) {
  if (!model || !model.modelName || !model.schema) {
    throw new Error('First arg for Resolver updateOne() should be instance of Mongoose Model.');
  }

  if (!tc || tc.constructor.name !== 'ObjectTypeComposer') {
    throw new Error('Second arg for Resolver updateOne() should be instance of ObjectTypeComposer.');
  }

  const findOneResolver = (0, _findOne.default)(model, tc, opts);
  const outputTypeName = `UpdateOne${tc.getTypeName()}Payload`;
  const outputType = tc.schemaComposer.getOrCreateOTC(outputTypeName, t => {
    t.addFields({
      recordId: {
        type: 'MongoID',
        description: 'Updated document ID'
      },
      record: {
        type: tc,
        description: 'Updated document'
      }
    });
  });
  const resolver = tc.schemaComposer.createResolver({
    name: 'updateOne',
    kind: 'mutation',
    description: 'Update one document: ' + '1) Retrieve one document via findOne. ' + '2) Apply updates to mongoose document. ' + '3) Mongoose applies defaults, setters, hooks and validation. ' + '4) And save it.',
    type: outputType,
    args: _objectSpread({}, (0, _helpers.recordHelperArgs)(tc, _objectSpread({
      recordTypeName: `UpdateOne${tc.getTypeName()}Input`,
      removeFields: ['id', '_id'],
      isRequired: true
    }, opts && opts.record)), {}, (0, _helpers.filterHelperArgs)(tc, model, _objectSpread({
      filterTypeName: `FilterUpdateOne${tc.getTypeName()}Input`,
      model
    }, opts && opts.filter)), {}, (0, _helpers.sortHelperArgs)(tc, model, _objectSpread({
      sortTypeName: `SortUpdateOne${tc.getTypeName()}Input`
    }, opts && opts.sort)), {}, (0, _helpers.skipHelperArgs)()),
    resolve: async resolveParams => {
      const recordData = resolveParams.args && resolveParams.args.record || null;
      const filterData = resolveParams.args && resolveParams.args.filter || {};

      if (!(typeof filterData === 'object') || Object.keys(filterData).length === 0) {
        return Promise.reject(new Error(`${tc.getTypeName()}.updateOne resolver requires at least one value in args.filter`));
      } // We should get all data for document, cause Mongoose model may have hooks/middlewares
      // which required some fields which not in graphql projection
      // So empty projection returns all fields.


      resolveParams.projection = {};
      let doc = await findOneResolver.resolve(resolveParams);

      if (resolveParams.beforeRecordMutate) {
        doc = await resolveParams.beforeRecordMutate(doc, resolveParams);
      }

      if (doc && recordData) {
        doc.set(recordData);
        await doc.save();
      }

      if (doc) {
        return {
          record: doc,
          recordId: tc.getRecordIdFn()(doc)
        };
      }

      return null;
    }
  });
  return resolver;
}
